pkg/optimizer/flatzinc_model.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020 IBM Corp.
+// SPDX-License-Identifier: Apache-2.0
+
+package optimizer
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"strings"
+)
+
+type FlatZincParam struct {
+	NumInstances uint // NumInstances > 1 means an array of params
+	Name         string
+	Type         string
+	Assignment   string
+}
+
+func (fzp FlatZincParam) ParamDeclaration() string {
+	if fzp.NumInstances == 1 {
+		return fmt.Sprintf("%s: %s = %s;\n", fzp.Type, fzp.Name, fzp.Assignment)
+	}
+	return fmt.Sprintf("array [1..%d] of %s: %s = %s;\n", fzp.NumInstances, fzp.Type, fzp.Name, fzp.Assignment)
+}
+
+type Annotations []string
+
+func (annots Annotations) AnnotationString() string {
+	annotsStr := strings.Join(annots, " :: ")
+	if len(annotsStr) > 0 {
+		annotsStr = " :: " + annotsStr
+	}
+	return annotsStr
+}
+
+type FlatZincVariable struct {
+	NumInstances uint // NumInstances > 1 means an array of vars
+	Name         string
+	Type         string
```

Comment:
This can not be a free string. It's better to define a type over the string and specify the valid values as constants including "int", "float", ...

Reply:
It's not that easy, because type can be something like `6..10` or `0.7..5.3` or `{9, 15, 22}`, which are all types we may have to use

Reply:
Ok